### PR TITLE
fix some memory leak bugs

### PR DESCRIPTION
**Pull request by wangzhengyong**
fix some memory leak bugs

### CODE REVIEW

**Comment by liupan1111** [comment]
please resolve the conflict first.

File: src/journal/JournalRecorder.cc [hunk 1]
```diff
@@ -130,6 +130,7 @@ void JournalRecorder::flush(Context *on_safe) {
 
   // avoid holding the lock in case there is nothing to flush
   ctx->complete(0);
```
Comment (liupan1111) [inline src/journal/JournalRecorder.cc]: complete(0) will delete itself

File: src/librbd/internal.cc [hunk 2]
```diff
@@ -609,15 +610,18 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
...
 	    imctx->state->close();
+            delete imctx;
 	    return r;
 	  }
 	}
 
 	r = imctx->state->close();
         if (r < 0) {
           lderr(cct) << "failed to close image: " << cpp_strerror(r) << dendl;
+          delete imctx;
```
Comment (runsisi) [inline src/librbd/internal.cc]: `ImageState<I>::close()` has deleted it.

File: src/librados/IoCtxImpl.cc [hunk 3]
```diff
@@ -1757,6 +1757,9 @@ int librados::IoCtxImpl::notify(const object_t& oid, bufferlist& bl,
   objecter->linger_cancel(linger_op);
 
   set_sync_op_version(objver);
+
+  delete notify_finish;
```
Comment (runsisi) [inline src/librados/IoCtxImpl.cc]: i think either `notify_finish->complete(r);` on line 1757 or `Objecter::handle_watch_notify` will delete it.

**Comment by gregsfortytwo** [comment]
As @runsisi said, Context objects are deleted by invoking complete() (and that happens in another thread in the branch where we Wait() on its Cond).

**Pull request closed without merging**